src/software/power/control_executor.cpp

```diff
@@ -0,0 +1,94 @@
+{
+}
+
+void ControlExecutor::execute(const TbotsProto_PowerControl& control)
+{
+    switch (control.chicker.which_chicker_command)
+    {
+        case TbotsProto_PowerControl_ChickerControl_kick_speed_m_per_s_tag:
+            detachInterrupt(digitalPinToInterrupt(BREAK_BEAM_PIN));
+            chicker->setKickSpeedMPerS(
+                control.chicker.chicker_command.kick_speed_m_per_s);
+            if (control.geneva.angle_deg != geneva->getCurrentAngle())
+            {
+                geneva->setAngle(control.geneva.angle_deg, &chicker->kick);
+            }
+            else
+            {
+                charger->setChargeDoneCallbackOnce(&chicker->kick);
+            }
+            charger->setChargeMode(HIGH);
+            break;
+        case TbotsProto_PowerControl_ChickerControl_chip_distance_meters_tag:
+            detachInterrupt(digitalPinToInterrupt(BREAK_BEAM_PIN));
+            chicker->setChipDistanceMeters(
+                control.chicker.chicker_command.chip_distance_meters);
+            if (control.geneva.angle_deg != geneva->getCurrentAngle())
+            {
+                geneva->setAngle(control.geneva.angle_deg, &chicker->chip);
+            }
+            else
+            {
+                charger->setChargeDoneCallbackOnce(&chicker->chip);
+            }
+            charger->setChargeMode(HIGH);
+            break;
+        case TbotsProto_PowerControl_ChickerControl_auto_chip_or_kick_tag:
+            detachInterrupt(digitalPinToInterrupt(BREAK_BEAM_PIN));
+            switch (
+                control.chicker.chicker_command.auto_chip_or_kick.which_auto_chip_or_kick)
+            {
+                case TbotsProto_AutoChipOrKick_autokick_speed_m_per_s_tag:
+                    chicker->setKickSpeedMPerS(
+                        control.chicker.chicker_command.kick_speed_m_per_s);
+                    if (control.geneva.angle_deg != geneva->getCurrentAngle())
+                    {
+                        geneva->setAngle(control.geneva.angle_deg, &chicker->autokick);
+                    }
+                    else
+                    {
+                        charger->setChargeDoneCallbackOnce(&chicker->autokick);
+                    }
+                    charger->setChargeMode(HIGH);
+                    break;
+                case TbotsProto_AutoChipOrKick_autochip_distance_meters_tag:
+                    chicker->setChipDistanceMeters(
+                        control.chicker.chicker_command.chip_distance_meters);
+                    if (control.geneva.angle_deg != geneva->getCurrentAngle())
+                    {
+                        geneva->setAngle(control.geneva.angle_deg, &chicker->autochip);
+                    }
+                    else
+                    {
+                        charger->setChargeDoneCallbackOnce(&chicker->autochip);
+                    }
+                    charger->setChargeMode(HIGH);
```

Comment:
yea thats currently what it does we register a callback (kick/chip) that gets triggered when charge done pin goes from low->high. So we want to have a delay between charges? that seems like it would affect gameplay

Reply:
We need to explicitly set chargeMode low after we kick I think, lets check with @ds570001

Reply:
At the moment I haven’t tried putting the flyback in regulation mode, so to keep the capacitors charged we will need to toggle the charge pin every so often if we want to keep the capacitors constantly charged at 238V. I think if we toggle it ~ every 5s then it will stay maximally charged. I will make this a priority to test out though, because if I can put the chip in regulation mode then you just need to hold charge high and the voltage will stay at 238V. However it would be nice to have this 5s toggle functionality implemented in case anything goes wrong with the regulation. If the regulation works, CHARGE just needs to be held high